src/DynamoCoreWpf/Extensions/ViewLoadedParams.cs

```diff
@@ -69,14 +69,23 @@ public Window DynamoWindow
             dynamoMenu = dynamoView.titleBar.ChildOfType<Menu>();
             viewStartupParams = new ViewStartupParams(dynamoVM);
             DynamoSelection.Instance.Selection.CollectionChanged += OnSelectionCollectionChanged;
-
         }
 
         public void AddMenuItem(MenuBarType type, MenuItem menuItem, int index = -1)
         {
             AddItemToMenu(type, menuItem, index);
         }
 
+        /// <summary>
+        /// Adds the extension UI control to a new tab in the right side bar.
+        /// </summary>
+        /// <param name="contentControl"></param>
+        /// <returns></returns>
+        internal void AddToExtensionsSideBar(ContentControl contentControl)
```

Comment:
I've added some thoughts below - heres a proposal for you and @QilongTang to consider:

instead of making clients call this API and add as many tabs as they want to this tabControl - each extension implements a method where they return a `contentControl` 
We call this method sometime after we call `loaded` on viewExtensions - (I actually think calling this API currently from loaded will result in weirdness)

We call it like `AddToExtensionsSideBar(extension.GetContent(),extension));`

This has some consequences and some constraints - 
* extensions can only inject themselves once (using this API atleast)
* if they miss this chance, their tabs cannot be added again later until Dynamo is restarted.
* if the extensions do not implement this method, we should not crash.

We need to think about the balance between flexibility and maintaining a logical experience.

I'm not sure this is a much better solution - but it is more constrained.

Reply:
like you said @reddyashish  - if we don't do any of the above, we can also have the client pass the extension anyway, and still have them implement a method we can call - this lets us change the implementation later and always have access to the extension name and other events so we can keep track of what extension has created what tab/content.

IE - we could make the API simply:

`AddExtensionContentToSideBar(ViewExtension)`

Reply:
Per our discussion, it seems we would like to 

- update it as `AddExtensionContentToSideBar(ContentControl ,ViewExtension)` so we could use extension GUID for tracking and use extension name for tab context
- give extension the right to inject anytime they want
- only create single tab for single extension

Feels like we need a wiki for this!

Reply:
Yes, made these changes. One issue was the extension GUID was different for each instance of the same extension so couldn't use that for tracking, so used the full class name of the extension to see if it is already added to the right side bar.